### PR TITLE
[lxc] initial integration

### PR DESCRIPTION
**Pull request by evverx**
It seems to be working in the sense that the fuzz target has found a bunch of issues like
```
==13==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f90db82553f at pc 0x00000067ebde bp 0x7ffcfe224790 sp 0x7ffcfe224788
READ of size 1 at 0x7f90db82553f thread T0
SCARINESS: 27 (1-byte-read-stack-buffer-overflow)
    #0 0x67ebdd in parse_byte_size_string /src/lxc/src/lxc/string_utils.c:926:24
    #1 0x426525 in set_config_console_buffer_size /src/lxc/src/lxc/confile.c:2436:8
    #2 0x5b2f04 in parse_line /src/lxc/src/lxc/confile.c:2953:9
    #3 0x448691 in lxc_file_for_each_line_mmap /src/lxc/src/lxc/parse.c:125:9
    #4 0x5b27e5 in lxc_config_read /src/lxc/src/lxc/confile.c:3024:9
    #5 0x581c03 in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:18:2
    #6 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
    #7 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
    #8 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
    #9 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
    #10 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
    #11 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #12 0x7f90da7c683f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
    #13 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
```
and
```
==31==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000268c0 at pc 0x0000005c4093 bp 0x7ffeaa8eb1b0 sp 0x7ffeaa8eb1a8
READ of size 8 at 0x6120000268c0 thread T0
SCARINESS: 51 (8-byte-read-heap-use-after-free)
    #0 0x5c4092 in lxc_get_netdev_by_idx /src/lxc/src/lxc/confile_utils.c:210:16
    #1 0x44024f in set_config_net_nic /src/lxc/src/lxc/confile.c:5261:11
    #2 0x5b2f04 in parse_line /src/lxc/src/lxc/confile.c:2953:9
    #3 0x448691 in lxc_file_for_each_line_mmap /src/lxc/src/lxc/parse.c:125:9
    #4 0x5b27e5 in lxc_config_read /src/lxc/src/lxc/confile.c:3024:9
    #5 0x581c03 in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:18:2
    #6 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
    #7 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
    #8 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
    #9 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
    #10 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
    #11 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #12 0x7fd8d480b83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
    #13 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
```

As far as I know, the lxc team is interested in integrating lxc into OSS-Fuzz.

cc @brauner

### CODE REVIEW

File: projects/lxc/fuzz-lxc-config-read.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	char tmpf[] = "fuzz-lxc-config-read-XXXXXX";
+	struct lxc_conf *conf = NULL;
+
+	fd = lxc_make_tmpfile(tmpf, false);
+	lxc_write_nointr(fd, data, size);
+	close(fd);
+
+	conf = lxc_conf_init();
+	lxc_config_read(tmpf, conf, false);
+	lxc_conf_free(conf);
```
Comment (evverx) [inline projects/lxc/fuzz-lxc-config-read.c:37]: @brauner this line triggers ASan from time to time with something like:
```
==13==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000266f0 at pc 0x0000005c7bfd bp 0x7ffe9a3f1570 sp 0x7ffe9a3f1568
READ of size 8 at 0x6120000266f0 thread T0
SCARINESS: 51 (8-byte-read-heap-use-after-free)
    #0 0x5c7bfc in lxc_free_netdev /src/lxc/src/lxc/confile_utils.c:414:15
    #1 0x5c7300 in lxc_free_networks /src/lxc/src/lxc/confile_utils.c:482:3
    #2 0x59cd2a in lxc_conf_free /src/lxc/src/lxc/conf.c:3961:2
    #3 0x581c0b in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:19:2
    #4 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
    #5 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
    #6 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
    #7 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
    #8 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
    #9 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #10 0x7f8900f9f83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
    #11 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
```
I'm not sure whether it's an lxc bug or a bug in the fuzz target. I kind of assumed that it's always safe to call `lxc_conf_free` right after `lxc_config_read` regardless of whether `lxc_config_read` has failed or not.

**Comment by brauner** [comment]
First, let me say _thank you so much_ for doing this! :) It's been on my TODO for some time but I never got around to actually hooking it up!

File: projects/lxc/fuzz-lxc-config-read.c (hunk 1 above)
Comment (brauner) [reply projects/lxc/fuzz-lxc-config-read.c:37]: This looks like a legitimate bug!
I sent a PR to fix this!

**Comment by brauner** [comment]
> It seems to be working in the sense that the fuzz target has found a bunch of issues like
> 
> ```
> ==13==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f90db82553f at pc 0x00000067ebde bp 0x7ffcfe224790 sp 0x7ffcfe224788
> READ of size 1 at 0x7f90db82553f thread T0
> SCARINESS: 27 (1-byte-read-stack-buffer-overflow)
>     #0 0x67ebdd in parse_byte_size_string /src/lxc/src/lxc/string_utils.c:926:24
>     #1 0x426525 in set_config_console_buffer_size /src/lxc/src/lxc/confile.c:2436:8
>     #2 0x5b2f04 in parse_line /src/lxc/src/lxc/confile.c:2953:9
>     #3 0x448691 in lxc_file_for_each_line_mmap /src/lxc/src/lxc/parse.c:125:9
>     #4 0x5b27e5 in lxc_config_read /src/lxc/src/lxc/confile.c:3024:9
>     #5 0x581c03 in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:18:2
>     #6 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
>     #7 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
>     #8 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
>     #9 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
>     #10 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
>     #11 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
>     #12 0x7f90da7c683f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
>     #13 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
> ```
> 
> and
> 
> ```
> ==31==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000268c0 at pc 0x0000005c4093 bp 0x7ffeaa8eb1b0 sp 0x7ffeaa8eb1a8
> READ of size 8 at 0x6120000268c0 thread T0
> SCARINESS: 51 (8-byte-read-heap-use-after-free)
>     #0 0x5c4092 in lxc_get_netdev_by_idx /src/lxc/src/lxc/confile_utils.c:210:16
>     #1 0x44024f in set_config_net_nic /src/lxc/src/lxc/confile.c:5261:11
>     #2 0x5b2f04 in parse_line /src/lxc/src/lxc/confile.c:2953:9
>     #3 0x448691 in lxc_file_for_each_line_mmap /src/lxc/src/lxc/parse.c:125:9
>     #4 0x5b27e5 in lxc_config_read /src/lxc/src/lxc/confile.c:3024:9
>     #5 0x581c03 in LLVMFuzzerTestOneInput /src/fuzz-lxc-config-read.c:18:2
>     #6 0x484783 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:599:15
>     #7 0x483c8a in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:505:3
>     #8 0x4859e7 in fuzzer::Fuzzer::MutateAndTestOne() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:745:19
>     #9 0x486575 in fuzzer::Fuzzer::Loop(std::__Fuzzer::vector<fuzzer::SizedFile, fuzzer::fuzzer_allocator<fuzzer::SizedFile> >&) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:883:5
>     #10 0x474307 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:906:6
>     #11 0x4a0162 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
>     #12 0x7fd8d480b83f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2083f)
>     #13 0x449278 in _start (/out/fuzz-lxc-config-read+0x449278)
> ```
> 
> As far as I know, the lxc team is interested in integrating lxc into OSS-Fuzz.
> 
> cc @brauner

Very much so! Thank you so much!

File: projects/lxc/project.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+homepage: "https://github.com/lxc/lxc"
+language: c
+primary_contact: ""
```
Comment (evverx) [inline projects/lxc/project.yaml:3]: @brauner I forgot to ask who OSS-Fuzz should send bug reports to: https://google.github.io/oss-fuzz/getting-started/new-project-guide/#primary. Unfortunately, it only works with gmail addresses: https://google.github.io/oss-fuzz/faq/#why-do-you-require-a-google-account-for-authentication

Comment (brauner) [reply projects/lxc/project.yaml:3]: Ah, I think:
christian.brauner@ubuntu.com
stgraber@ubuntu.com
they are both aliases and should be gmail. If that doesn't work:
stephane.graber@canonical.com
christian.brauner@canonical.com

File: projects/lxc/fuzz-lxc-config-read.c [hunk 3]
```diff
@@ -0,0 +1,41 @@
+/*
+# Copyright 2021 Google Inc.
```
Comment (jonathanmetzman) [inline projects/lxc/fuzz-lxc-config-read.c]: Lol at just copying the python/docker license header into a C comment.

**Review by jonathanmetzman — APPROVED** [review]
LGTM

**Pull request merged**